### PR TITLE
Make debug box statically configurable

### PR DESCRIPTION
**Pull request by danny4478**
Modifications:
- SVC for debug box registration removed
- Registration to debug_box is now done statically
- Documentation updated

Testing:
- Tested on public box (debug-fault example) - OK
- Tested on secure box (number-store example) - OK
- Run on uvisor-tests - OK

### CODE REVIEW

File: api/inc/box_config.h [hunk 1]
```diff
@@ -79,7 +88,7 @@ UVISOR_EXTERN void const * const public_box_cfg_ptr;
 /* this macro selects an overloaded macro (variable number of arguments) */
 #define __UVISOR_BOX_MACRO(_1, _2, _3, _4, NAME, ...) NAME
 
-#define __UVISOR_BOX_CONFIG(box_name, acl_list, acl_list_count, stack_size, context_size) \
+#define __UVISOR_BOX_CONFIG(box_name, acl_list, acl_list_count, stack_size, context_size, debug_box_ptr) \
```
Comment (Patater) [inline api/inc/box_config.h]: Please add `debug_box_ptr` to the uVisor config instead of the box config, since each box doesn't need their own debug box. The system as a whole will have one debug box, and this is best set in one spot: the uVisor config.

You'll be able to get away without modifying so many macros this way. You can use `g_debug_driver` directly from `uvisor-input.S`.

File: api/inc/debug_exports.h [hunk 2]
```diff
@@ -19,11 +19,17 @@
 
 #include <stdint.h>
 
+
+#define UVISOR_DEBUG_BOX_MAGIC      (0xDB4807DB)
```
Comment (Patater) [inline api/inc/debug_exports.h]: Please place this magic with all the other magic. Also, generate the magic in the same way as the other magics.

File: core/vmpu/src/vmpu.c [hunk 3]
```diff
@@ -431,6 +463,13 @@ static void vmpu_enumerate_boxes(void)
         box_init(index, box_cfgtbl);
     }
 
+    if(debug_box_validated) {
```
Comment (Patater) [inline core/vmpu/src/vmpu.c]: Place a space between keywords like `if` and `(`.

File: docs/lib/DEBUGGING.md [hunk 4]
```diff
@@ -121,16 +121,17 @@ The uVisor code is instrumented to output debug information when it is relevant.
 
 Instead, debug events and messages are forwarded to a special unprivileged box, called a *debug box*. A debug box is configured just like any other secure box, but it registers with uVisor to handle debug callbacks. These callbacks must adhere to a format that uVisor provides, in the form of a debug box driver.
 
-The debug box driver is encoded in a standard table (a C `struct`) that a debug box must populate at initialization time. A debug box can decide to implement only some of the available handlers though they must all exist at least as empty functions. Otherwise, the program behavior might be unpredictable.
+The debug box driver is encoded in a standard table (a C `struct`) that a debug box must populate statically. A debug box can decide to implement only some of the available handlers though they must all exist at least as empty functions. Otherwise, the program behavior might be unpredictable.
```
Comment (Patater) [inline docs/lib/DEBUGGING.md]: Very good to see the documentation updated as well. Thanks.

File: api/inc/box_config.h [hunk 5]
```diff
@@ -169,4 +191,15 @@ UVISOR_EXTERN void const * const public_box_cfg_ptr;
 
 #define __uvisor_ctx (((UvisorBoxIndex *) __uvisor_ps)->bss.address_of.context)
 
+
+/* Use this macro before trying to configure a box with debug_box enabled.
+ * It will create a valid debug driver struct with the halt_error_func parameter as its halt_error() function */
+#define UVISOR_GENERATE_DEBUG_DRIVER(halt_error_func) \
```
Comment (Patater) [inline api/inc/box_config.h]: Consider renaming to `UVISOR_DEBUG_BOX`

File: core/vmpu/src/vmpu.c [hunk 6]
```diff
@@ -414,6 +434,18 @@ static void vmpu_enumerate_boxes(void)
         int index = box_order[box_id];
         UvisorBoxConfig const * box_cfgtbl = ((UvisorBoxConfig const * *) __uvisor_config.cfgtbl_ptr_start)[index];
 
+        if(box_cfgtbl->debug_box_ptr != NULL) {
+            if(++debug_box_count > MAX_DEBUG_BOX_COUNT) {
+                HALT_ERROR(SANITY_CHECK_FAILED, "More than one box is registered to the debug box!\n");
```
Comment (Patater) [inline core/vmpu/src/vmpu.c]: This won't be possible when we use uVisor config, since there will be multiple definition errors at build time if multiple debug boxes are created.

File: api/inc/debug_exports.h [hunk 7]
```diff
@@ -18,12 +18,19 @@
...
+
+
+/* Debug box driver -- Version 1
  * A constant instance of this struct must be instantiated by the unprivileged
  * code to setup a debug box.*/
 typedef struct TUvisorDebugDriver {
-    uint32_t (*get_version)(void);
+    const uint32_t magic;
+    const uint32_t version;
+    const UvisorBoxConfig * box_cfg_ptr;
```
Comment (alzix) [inline api/inc/debug_exports.h]: please add another const. the more the merrier

File: api/src/uvisor-input.S [hunk 8]
```diff
@@ -122,6 +126,11 @@ __uvisor_page_size:
      * The user can override this weak reference. */
     .long 16384
 
+__uvisor_debug_driver_ptr:
+    /* No debug driver by default.
+    * Populated per request by the application */
```
Comment (alzix) [inline api/src/uvisor-input.S]: please align multi-line comments by asterisks. add one more space

File: docs/lib/DEBUGGING.md [hunk 9]
```diff
@@ -165,17 +155,16 @@ static void halt_error(int reason) {
...
 }
 
-static void box_debug_main(const void *)
-{
-    /* Debug box driver -- Version 0 */
-    static const TUvisorDebugDriver driver = {
-        get_version,
-        halt_error
-    };
+/* Create a valid debug driver struct and use example_halt_error() as its halt_error() function */
```
Comment (alzix) [inline docs/lib/DEBUGGING.md]: i believe we should delete this code snippet and add a a reference to our https://github.com/ARMmbed/mbed-os-example-uvisor-debug-fault.
perhaps it is already done by Oren's commit...

Comment (Patater) [inline docs/lib/DEBUGGING.md]: It's in the scope of Oren's work, yes. I don't believe it has been done by any PR from Oren yet, though.

**Comment by danny4478** [comment]
This PR is ready for review.
Last commit points to corresponding PR in uvisor_tests.

File: core/vmpu/src/vmpu.c [hunk 10]
```diff
@@ -397,6 +401,8 @@ static void vmpu_configure_box_sram(uint8_t box_id, UvisorBoxConfig const * box_
 
 static void vmpu_enumerate_boxes(void)
 {
+    bool debug_box_validated = FALSE;
```
Comment (danny4478) [inline core/vmpu/src/vmpu.c]: Done

File: core/vmpu/src/vmpu.c [hunk 11]
```diff
@@ -413,6 +419,32 @@ static void vmpu_enumerate_boxes(void)
         /* Select the pointer to the (permuted) box configuration table. */
         int index = box_order[box_id];
         UvisorBoxConfig const * box_cfgtbl = ((UvisorBoxConfig const * *) __uvisor_config.cfgtbl_ptr_start)[index];
+        TUvisorDebugDriver const * const debug_driver = __uvisor_config.debug_driver;
+
+        /* If a debug driver was created by the application and it is associated to this box_id,
+         * initialize g_debug_box accordingly.
+         * Note: The verification that no more than one box was created and been associated with the debug driver
+         * is done statically on debug driver creation time [in build time] */
```
Comment (danny4478) [inline core/vmpu/src/vmpu.c]: I think the correct assert is: assert(!debug_box_validated);

File: core/vmpu/src/vmpu.c [hunk 12]
```diff
@@ -413,6 +419,32 @@ static void vmpu_enumerate_boxes(void)
...
+        if (debug_driver != NULL) {
+            if (debug_driver->box_cfg_ptr == box_cfgtbl) {
```
Comment (danny4478) [inline core/vmpu/src/vmpu.c]: Done

File: core/vmpu/src/vmpu.c (hunk 3 above)
Comment (danny4478) [inline core/vmpu/src/vmpu.c]: They have different meanings.
- g_debug_box.initialized means the debug driver is ready to be used. 
- debug_box_validated is a flag denoting that the debug driver is valid and should be set as initialized LATER, after all initialization(more specifically, g_debug_interrupt_sp[] initialization) is done.

File: core/vmpu/src/vmpu.c [hunk 13]
```diff
@@ -431,6 +463,13 @@ static void vmpu_enumerate_boxes(void)
         box_init(index, box_cfgtbl);
     }
 
+    if(debug_box_validated) {
+        for (int ii = 0; ii < UVISOR_MAX_BOXES; ii++) {
+            g_debug_interrupt_sp[ii] = g_context_current_states[ii].sp;
```
Comment (danny4478) [inline core/vmpu/src/vmpu.c]: 1. This assignment cannot be done before g_debug_interrupt_sp[] is being initialized inside vmpu_configure_box_sram().
2. I agree. But this is not related to the PR. This can / should be done in another PR.

File: docs/lib/DEBUGGING.md [hunk 14]
```diff
@@ -121,9 +121,9 @@ The uVisor code is instrumented to output debug information when it is relevant.
 
 Instead, debug events and messages are forwarded to a special unprivileged box, called a *debug box*. A debug box is configured just like any other secure box, but it registers with uVisor to handle debug callbacks. These callbacks must adhere to a format that uVisor provides, in the form of a debug box driver.
 
-The debug box driver is encoded in a standard table (a C `struct`) that a debug box must populate at initialization time. A debug box can decide to implement only some of the available handlers though they must all exist at least as empty functions. Otherwise, the program behavior might be unpredictable.
+The debug box driver is encoded in a standard table (a C `struct`) that a debug box must populate statically. A debug box can decide to implement only some of the available handlers though they must all exist at least as empty functions. Otherwise, the program behavior might be unpredictable.
```
Comment (danny4478) [inline docs/lib/DEBUGGING.md]: Originally this document was VERY specific, with code examples.
But, latest PRs removed the code examples, referring the reader to the the debug-fault example.

Comment (danny4478) [inline docs/lib/DEBUGGING.md]: @Patater - Your thoughts?

File: core/vmpu/src/vmpu.c (hunk 3 above)
Comment (danny4478) [inline core/vmpu/src/vmpu.c]: Done

File: core/vmpu/src/vmpu.c (hunk 11 above)
Comment (alzix) [inline core/vmpu/src/vmpu.c]: done

File: core/vmpu/src/vmpu.c (hunk 3 above)
Comment (alzix) [inline core/vmpu/src/vmpu.c]: dropped - shall be handled in dedicated PR as it changes original logic

File: core/vmpu/src/vmpu.c [hunk 15]
```diff
@@ -27,6 +29,8 @@
 #include "vmpu_unpriv_access.h"
 #include <sys/reent.h>
 
+
+uint32_t g_debug_interrupt_sp[UVISOR_MAX_BOXES];
```
Comment (Patater) [inline core/vmpu/src/vmpu.c]: We have another of `g_debug_interrupt_sp` in `core/debug/src/debug_box.c`. Why doesn't this cause multiple definition linker errors?

File: core/vmpu/src/vmpu.c [hunk 16]
```diff
@@ -413,6 +419,33 @@ static void vmpu_enumerate_boxes(void)
         /* Select the pointer to the (permuted) box configuration table. */
         int index = box_order[box_id];
         UvisorBoxConfig const * box_cfgtbl = ((UvisorBoxConfig const * *) __uvisor_config.cfgtbl_ptr_start)[index];
+        TUvisorDebugDriver const * const debug_driver = __uvisor_config.debug_driver;
+
+        /* If a debug driver was created by the application and it is associated to this box_id,
+         * initialize g_debug_box accordingly.
+         * Note: The verification that no more than one box was created and been associated with the debug driver
+         * is done statically on debug driver creation time [in build time] */
+        if ((debug_driver != NULL) && (debug_driver->box_cfg_ptr == box_cfgtbl)) {
```
Comment (Patater) [inline core/vmpu/src/vmpu.c]: `debug_driver` should never be `NULL`. Consider changing to an `assert` or `HALT_ERROR(SANITY_CHECKFAILED`. It's a programmer error if `debug_driver` is `NULL`.

File: docs/lib/DEBUGGING.md (hunk 14 above)
Comment (Patater) [inline docs/lib/DEBUGGING.md]: Danny's changes to the documentation here are good.

File: core/vmpu/src/vmpu.c (hunk 16 above)
Comment (danny4478) [inline core/vmpu/src/vmpu.c]: Done

File: core/vmpu/src/vmpu.c (hunk 15 above)
Comment (danny4478) [inline core/vmpu/src/vmpu.c]: Apparently it is. It has got there probably on the latest merge from Roman's branch...
We do not get any errors.
Should check why.

Comment (danny4478) [inline core/vmpu/src/vmpu.c]: It has got there probably on the latest merge from Roman's branch...
We do not get any errors.
Should check why.

File: api/src/uvisor-input.S [hunk 17]
```diff
@@ -122,6 +126,14 @@ __uvisor_page_size:
      * The user can override this weak reference. */
     .long 16384
 
+__uvisor_debug_driver:
+    /* No debug driver by default.
+    * Populated per request by the application */
```
Comment (Patater) [inline api/src/uvisor-input.S]: Please align multi-line comments by asterisks. Add one more space.

Comment (danny4478) [inline api/src/uvisor-input.S]: Done

File: api/inc/box_config.h [hunk 18]
```diff
@@ -169,4 +170,21 @@ UVISOR_EXTERN void const * const public_box_cfg_ptr;
...
+#define UVISOR_DEBUG_DRIVER(box_name, halt_error_func) \
+        UVISOR_EXTERN TUvisorDebugDriver const __uvisor_debug_driver; \
+        TUvisorDebugDriver const __uvisor_debug_driver = { \
+            UVISOR_DEBUG_BOX_MAGIC, \
+            UVISOR_DEBUG_BOX_VERSION, \
+            &box_name ## _cfg, \
+            halt_error_func \
+        };
+
+/* Use this macro after calling the box configuration macro, in order to register the public box as a debug box. */
```
Comment (Patater) [inline api/inc/box_config.h]: Ensure this line is wrapped to 79 columns.

File: api/inc/debug_exports.h [hunk 19]
```diff
@@ -19,12 +19,19 @@
...
 #include "api/inc/halt_exports.h"
 #include <stdint.h>
+#include "api/inc/vmpu_exports.h"
 
-/* Debug box driver -- Version 0
+
+#define UVISOR_DEBUG_BOX_VERSION    (1)
+
+
+/* Debug box driver -- Version 1
```
Comment (Patater) [inline api/inc/debug_exports.h]: Remove ` -- Version 1`. Let the `UVISOR_DEBUG_BOX_VERSION` be the only source of version, so it's easy to update.

File: core/debug/inc/debug.h [hunk 20]
```diff
@@ -39,7 +39,6 @@ void debug_sau_config(void);
 void debug_fault(THaltError reason, uint32_t lr, uint32_t sp);
 
 /* Debug box */
-void debug_register_driver(const TUvisorDebugDriver * const driver);
 uint32_t debug_get_version(void);
```
Comment (Patater) [inline core/debug/inc/debug.h]: Please remove debug_get_version and its implementation.
